File: Raven.Client.Lightweight/Changes/IDatabaseChanges.cs
using System;
using System.Collections.Concurrent;
using System.Threading.Tasks;
using Raven.Abstractions.Data;

namespace Raven.Client.Changes
{
    public interface IDatabaseChanges : IConnectableChanges<IDatabaseChanges>
	{
		IObservableWithTask<IndexChangeNotification> ForIndex(string indexName);
		IObservableWithTask<DocumentChangeNotification> ForDocument(string docId);
		IObservableWithTask<DocumentChangeNotification> ForAllDocuments();
		IObservableWithTask<IndexChangeNotification> ForAllIndexes();
	    IObservableWithTask<TransformerChangeNotification> ForAllTransformers();
        IObservableWithTask<DocumentChangeNotification> ForDocumentsStartingWith(string docIdPrefix);
		IObservableWithTask<DocumentChangeNotification> ForDocumentsInCollection(string collectionName);
		IObservableWithTask<DocumentChangeNotification> ForDocumentsInCollection<TEntity>();
		IObservableWithTask<DocumentChangeNotification> ForDocumentsOfType(string typeName);
		IObservableWithTask<DocumentChangeNotification> ForDocumentsOfType(Type type);
		IObservableWithTask<DocumentChangeNotification> ForDocumentsOfType<TEntity>();
		IObservableWithTask<ReplicationConflictNotification> ForAllReplicationConflicts();
		IObservableWithTask<BulkInsertChangeNotification> ForBulkInsert(Guid operationId);
	}
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/changes/IDatabaseChanges.java
package net.ravendb.client.changes;

import java.util.UUID;

import net.ravendb.abstractions.basic.EventHandler;
import net.ravendb.abstractions.basic.VoidArgs;
import net.ravendb.abstractions.data.BulkInsertChangeNotification;
import net.ravendb.abstractions.data.DocumentChangeNotification;
import net.ravendb.abstractions.data.IndexChangeNotification;
import net.ravendb.abstractions.data.ReplicationConflictNotification;


public interface IDatabaseChanges {

  public boolean isConnected();

  public void addConnectionStatusChanged(EventHandler<VoidArgs> handler);

  public void removeConnectionStatusChanges(EventHandler<VoidArgs> handler);

  public IObservable<IndexChangeNotification> forIndex(String indexName);

  public IObservable<DocumentChangeNotification> forDocument(String docId);

  public IObservable<DocumentChangeNotification> forAllDocuments();

  public IObservable<IndexChangeNotification> forAllIndexes();

  public IObservable<DocumentChangeNotification> forDocumentsStartingWith(String docIdPrefix);

  public IObservable<DocumentChangeNotification> forDocumentsInCollection(String collectionName);

  public IObservable<DocumentChangeNotification> forDocumentsInCollection(Class<?> clazz);

  public IObservable<DocumentChangeNotification> forDocumentsOfType(String typeName);

  public IObservable<DocumentChangeNotification> forDocumentsOfType(Class<?> clazz);

  public IObservable<ReplicationConflictNotification> forAllReplicationConflicts();

  public IObservable<BulkInsertChangeNotification> forBulkInsert(UUID operationId);

  void waitForAllPendingSubscriptions();
}
